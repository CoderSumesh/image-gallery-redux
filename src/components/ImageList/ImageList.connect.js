import { connect } from "react-redux";

import { fetchImages } from "../../store/actions";

import ImageList from "./ImageList";

const mapStateToProps = (state) => {
  const { isLoading, images, searchText } = state;
  return { isLoading, images, searchText };
};

const mapDispatchToProps = (dispatch) => ({
  fetchImages: () => dispatch(fetchImages()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ImageList);
