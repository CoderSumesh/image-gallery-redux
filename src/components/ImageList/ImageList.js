import React, { useEffect, useState } from "react";
import { connect } from "react-redux";

import { ImageCard } from "../ImageCard";

import {
  SORT_MOST_LIKED,
  SORT_MOST_COMMENTED,
  fetchImages,
} from "../../store/actions";

function ImageList(props) {
  const { isLoading, images, fetchImages, sortType, searchText } = props;

  const [imageList, setImageList] = useState(images);

  useEffect(() => {
    fetchImages();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    let data = [...images];

    if (searchText) {
      data = data.filter((item) =>
        item.category.toLowerCase().includes(searchText.toLowerCase())
      );
    }

    setImageList([...data]);
  }, [images, searchText]);

  return (
    <div className="grid-container">
      {isLoading ? (
        <span className="spinner">Loading...</span>
      ) : (
        imageList.map((item) => (
          <div key={item.id} className="grid-item">
            <ImageCard data={{ ...item }} />
          </div>
        ))
      )}
    </div>
  );
}

const mapStateToProps = (state) => {
  const { isLoading, images, sortType, searchText } = state;
  return { isLoading, images, sortType, searchText };
};

const mapDispatchToProps = (dispatch) => ({
  fetchImages: () => dispatch(fetchImages()),
});

export default connect(mapStateToProps, mapDispatchToProps)(ImageList);
