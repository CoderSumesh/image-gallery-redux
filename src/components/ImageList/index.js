export {default as ImageList} from './ImageList.connect'