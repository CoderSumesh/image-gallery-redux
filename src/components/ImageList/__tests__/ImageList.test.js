
import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { createWaitForElement } from 'enzyme-wait';
import { Provider } from 'react-redux';
import moxios from 'moxios';

import Store from '../../../store';
import {ImageList} from '../ImageList';
import data from '../../../store/data';

describe('ImageList', () => {
    let useEffect;
    let component;
    let fetchProducts = jest.fn();
    let sortType = jest.fn();
    let searchText = jest.fn();
    let props = {
        isLoading: false,
        images: [],
        fetchProducts,
        sortType,
        searchText
    }

    beforeEach(() => {
        useEffect = jest.spyOn(React, "useEffect").mockImplementation(f => f());
        component = shallow(<Store><ImageList /></Store>)
    });

    it('renders properly', () => {
        expect(shallowToJson(component)).toMatchSnapshot();
    });
});

describe('ImageList', () => {
    let component, waitForListItem;

    beforeEach(() => {
        moxios.install();
        moxios.stubRequest('https://raw.githubusercontent.com/Lokenath/MyRepo/master/Test/package.json', {
            status: 200,
            response: data,
        });

        component = mount(<Store><ImageList /></Store>)
        waitForListItem = createWaitForElement('.card');
    });

    afterEach(() => {
        moxios.uninstall();
    });

    it("must render a loading span before api call success", () => {
        expect(component.find("span.spinner").exists()).toBeTruthy();
    });

    it('shows list items after some time', async done => {
        const componentReady = await waitForListItem(component);
        expect(componentReady.find('.card')).toHaveLength(data.length);
        done();
    })

});