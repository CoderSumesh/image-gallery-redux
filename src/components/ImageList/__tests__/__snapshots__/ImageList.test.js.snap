// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImageList renders properly 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Component />
</Provider>
`;

exports[`ProductList renders properly 1`] = `null`;
